sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveMetastoreCatalog.scala

```diff
@@ -217,6 +235,62 @@ private[hive] class HiveMetastoreCatalog(sparkSession: SparkSession) extends Log
     result.copy(expectedOutputAttributes = Some(relation.output))
   }
 
+  private def inferIfNeeded(
+      relation: CatalogRelation,
+      options: Map[String, String],
+      fallbackSchema: StructType,
+      fileFormat: FileFormat,
+      fileIndexOpt: Option[FileIndex] = None): (StructType, CatalogTable) = {
+    val inferenceMode = sparkSession.sessionState.conf.caseSensitiveInferenceMode
+    val shouldInfer = (inferenceMode != NEVER_INFER) && !relation.tableMeta.schemaPreservesCase
+    val tableName = relation.tableMeta.identifier.table
+    if (shouldInfer) {
+      logInfo(s"Inferring case-sensitive schema for table $tableName (inference mode: " +
+        s"$inferenceMode)")
+      val fileIndex = fileIndexOpt.getOrElse {
+        val rootPath = new Path(new URI(relation.tableMeta.location))
+        new InMemoryFileIndex(sparkSession, Seq(rootPath), options, None)
+      }
+
+      val inferredSchema = {
+        val schema = fileFormat.inferSchema(
+          sparkSession,
+          options,
+          fileIndex.listFiles(Nil).flatMap(_.files))
+        fileFormat match {
+          case _: ParquetFileFormat =>
+            schema.map(ParquetFileFormat.mergeMetastoreParquetSchema(relation.tableMeta.schema, _))
```

Comment:
This was used in the previous code for schema inference. I think the only reason this would still be needed is if the metastore schema contains a nullable field that isn't actually present in the underlying Parquet data. See #5214 for more details here.